File: packages/device/src/__tests__/DeviceModule.test.ts
import { describe, it, expect, beforeEach } from 'vitest';
import { RynBridge, MockTransport } from '@rynbridge/core';
import type { BridgeResponse } from '@rynbridge/core';
import { DeviceModule } from '../DeviceModule.js';

describe('DeviceModule', () => {
  let transport: MockTransport;
  let bridge: RynBridge;
  let device: DeviceModule;

  beforeEach(() => {
    transport = new MockTransport();
    bridge = new RynBridge({ timeout: 5000 }, transport);
    device = new DeviceModule(bridge);
  });

  function respondSuccess(payload: Record<string, unknown> = {}) {
    const sent = JSON.parse(transport.sent[transport.sent.length - 1]);
    const response: BridgeResponse = {
      id: sent.id,
      status: 'success',
      payload,
      error: null,
    };
    transport.simulateIncoming(JSON.stringify(response));
  }

  function simulateNativeEvent(module: string, action: string, payload: Record<string, unknown>) {
    const event = JSON.stringify({
      id: 'event-' + Math.random().toString(36).slice(2),
      module,
      action,
      payload,
      version: '1.0.0',
    });
    transport.simulateIncoming(event);
  }

  describe('getInfo', () => {
    it('sends correct module and action', async () => {
      const promise = device.getInfo();

      const sent = JSON.parse(transport.sent[0]);
      expect(sent.module).toBe('device');
      expect(sent.action).toBe('getInfo');
      expect(sent.payload).toEqual({});

      respondSuccess({ platform: 'ios', osVersion: '17.0', model: 'iPhone 15', appVersion: '1.0.0' });

      const result = await promise;
      expect(result).toEqual({
        platform: 'ios',
        osVersion: '17.0',
        model: 'iPhone 15',
        appVersion: '1.0.0',
      });
    });
  });

  describe('getBattery', () => {
    it('sends correct module and action', async () => {
      const promise = device.getBattery();

      const sent = JSON.parse(transport.sent[0]);
      expect(sent.module).toBe('device');
      expect(sent.action).toBe('getBattery');

      respondSuccess({ level: 85, isCharging: true });

      const result = await promise;
      expect(result).toEqual({ level: 85, isCharging: true });
    });
  });

  describe('getScreen', () => {
    it('sends correct module and action', async () => {
      const promise = device.getScreen();

      const sent = JSON.parse(transport.sent[0]);
      expect(sent.module).toBe('device');
      expect(sent.action).toBe('getScreen');

      respondSuccess({ width: 390, height: 844, scale: 3, orientation: 'portrait' });

      const result = await promise;
      expect(result).toEqual({ width: 390, height: 844, scale: 3, orientation: 'portrait' });
    });
  });

  describe('vibrate', () => {
    it('sends fire-and-forget without payload', () => {
      device.vibrate();

      expect(transport.sent).toHaveLength(1);
      const sent = JSON.parse(transport.sent[0]);
      expect(sent.module).toBe('device');
      expect(sent.action).toBe('vibrate');
      expect(sent.payload).toEqual({});
    });

    it('sends fire-and-forget with pattern', () => {
      device.vibrate({ pattern: [100, 50, 100] });

      const sent = JSON.parse(transport.sent[0]);
      expect(sent.module).toBe('device');
      expect(sent.action).toBe('vibrate');
      expect(sent.payload).toEqual({ pattern: [100, 50, 100] });
    });
  });

  describe('capturePhoto', () => {
    it('sends correct module and action with default payload', async () => {
      const promise = device.capturePhoto();

      const sent = JSON.parse(transport.sent[0]);
      expect(sent.module).toBe('device');
      expect(sent.action).toBe('capturePhoto');
      expect(sent.payload).toEqual({});

      respondSuccess({ imageBase64: 'abc123', width: 1920, height: 1080 });

      const result = await promise;
      expect(result).toEqual({ imageBase64: 'abc123', width: 1920, height: 1080 });
    });

    it('sends quality and camera options', async () => {
      const promise = device.capturePhoto({ quality: 0.8, camera: 'front' });

      const sent = JSON.parse(transport.sent[0]);
      expect(sent.payload).toEqual({ quality: 0.8, camera: 'front' });

      respondSuccess({ imageBase64: 'xyz', width: 640, height: 480 });

      const result = await promise;
      expect(result).toEqual({ imageBase64: 'xyz', width: 640, height: 480 });
    });
  });

  describe('getLocation', () => {
    it('returns location info', async () => {
      const promise = device.getLocation();

      const sent = JSON.parse(transport.sent[0]);
      expect(sent.module).toBe('device');
      expect(sent.action).toBe('getLocation');
      expect(sent.payload).toEqual({});

      respondSuccess({ latitude: 37.5665, longitude: 126.978, accuracy: 10 });

      const result = await promise;
      expect(result).toEqual({ latitude: 37.5665, longitude: 126.978, accuracy: 10 });
    });
  });

  describe('authenticate', () => {
    it('sends reason and returns success', async () => {
      const promise = device.authenticate({ reason: 'Confirm payment' });

      const sent = JSON.parse(transport.sent[0]);
      expect(sent.module).toBe('device');
      expect(sent.action).toBe('authenticate');
      expect(sent.payload).toEqual({ reason: 'Confirm payment' });

      respondSuccess({ success: true });

      const result = await promise;
      expect(result).toEqual({ success: true });
    });

    it('returns failure on denied authentication', async () => {
      const promise = device.authenticate({ reason: 'Login' });
      respondSuccess({ success: false });

      const result = await promise;
      expect(result).toEqual({ success: false });
    });
  });

  describe('onLocationChange', () => {
    it('subscribes to location events and returns unsubscribe function', () => {
      const received: unknown[] = [];
      const unsubscribe = device.onLocationChange((data) => {
        received.push(data);
      });

      simulateNativeEvent('device', 'locationChange', { latitude: 37.5, longitude: 127.0, accuracy: 5 });

      expect(received).toHaveLength(1);
      expect(received[0]).toEqual({ latitude: 37.5, longitude: 127.0, accuracy: 5 });

      unsubscribe();
      simulateNativeEvent('device', 'locationChange', { latitude: 38.0, longitude: 128.0, accuracy: 3 });

      expect(received).toHaveLength(1);
    });
  });

  describe('onBatteryChange', () => {
    it('subscribes to battery events and returns unsubscribe function', () => {
      const received: unknown[] = [];
      const unsubscribe = device.onBatteryChange((data) => {
        received.push(data);
      });

      simulateNativeEvent('device', 'batteryChange', { level: 75, isCharging: false });

      expect(received).toHaveLength(1);
      expect(received[0]).toEqual({ level: 75, isCharging: false });

      unsubscribe();
      simulateNativeEvent('device', 'batteryChange', { level: 50, isCharging: true });

      expect(received).toHaveLength(1);
    });
  });

  describe('error handling', () => {
    it('propagates bridge errors', async () => {
      const promise = device.getInfo();

      const sent = JSON.parse(transport.sent[0]);
      const response: BridgeResponse = {
        id: sent.id,
        status: 'error',
        payload: {},
        error: { code: 'UNKNOWN', message: 'Device info unavailable' },
      };
      transport.simulateIncoming(JSON.stringify(response));

      await expect(promise).rejects.toThrow('Device info unavailable');
    });
  });
});

File: android/device/src/test/kotlin/io/rynbridge/device/DeviceModuleTest.kt
package io.rynbridge.device

import io.rynbridge.core.*
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.delay
import kotlinx.coroutines.test.runTest
import kotlinx.coroutines.withContext
import kotlinx.serialization.json.Json
import org.junit.jupiter.api.Assertions.*
import org.junit.jupiter.api.Test

class DeviceModuleTest {

    @Test
    fun `getInfo returns device info`() = runTest {
        val provider = MockDeviceInfoProvider()
        val module = DeviceModule(provider)
        val handler = module.actions["getInfo"]!!

        val result = handler(emptyMap())
        assertEquals("android", result["platform"]?.stringValue)
        assertEquals("14", result["osVersion"]?.stringValue)
        assertEquals("Pixel 8", result["model"]?.stringValue)
        assertEquals("1.0.0", result["appVersion"]?.stringValue)
    }

    @Test
    fun `getBattery returns battery info`() = runTest {
        val provider = MockDeviceInfoProvider()
        val module = DeviceModule(provider)
        val handler = module.actions["getBattery"]!!

        val result = handler(emptyMap())
        assertEquals(85L, result["level"]?.intValue)
        assertEquals(true, result["isCharging"]?.boolValue)
    }

    @Test
    fun `getScreen returns screen info`() = runTest {
        val provider = MockDeviceInfoProvider()
        val module = DeviceModule(provider)
        val handler = module.actions["getScreen"]!!

        val result = handler(emptyMap())
        assertEquals(390.0, result["width"]?.doubleValue)
        assertEquals(844.0, result["height"]?.doubleValue)
        assertEquals(3.0, result["scale"]?.doubleValue)
        assertEquals("portrait", result["orientation"]?.stringValue)
    }

    @Test
    fun `vibrate with pattern`() = runTest {
        val provider = MockDeviceInfoProvider()
        val module = DeviceModule(provider)
        val handler = module.actions["vibrate"]!!

        val result = handler(mapOf("pattern" to BridgeValue.array(listOf(BridgeValue.int(100), BridgeValue.int(200)))))
        assertTrue(result.isEmpty())
        assertEquals(listOf(100, 200), provider.lastVibratePattern)
    }

    @Test
    fun `vibrate without pattern`() = runTest {
        val provider = MockDeviceInfoProvider()
        val module = DeviceModule(provider)
        val handler = module.actions["vibrate"]!!

        handler(emptyMap())
        assertEquals(emptyList<Int>(), provider.lastVibratePattern)
    }

    @Test
    fun `module name and version`() {
        val provider = MockDeviceInfoProvider()
        val module = DeviceModule(provider)
        assertEquals("device", module.name)
        assertEquals("0.1.0", module.version)
    }

    @Test
    fun `end to end with bridge`() = runTest {
        val transport = MockTransport()
        val bridge = RynBridge(transport = transport, config = BridgeConfig(timeout = 5000L))
        val provider = MockDeviceInfoProvider()
        bridge.register(DeviceModule(provider))

        val requestJSON = """{"id":"req-1","module":"device","action":"getInfo","payload":{},"version":"0.1.0"}"""
        transport.simulateIncoming(requestJSON)

        withContext(Dispatchers.Default) { delay(200) }

        assertEquals(1, transport.sent.size)
        val json = Json { ignoreUnknownKeys = true }
        val response = json.decodeFromString<BridgeResponse>(transport.sent[0])
        assertEquals("req-1", response.id)
        assertEquals(ResponseStatus.success, response.status)
        assertEquals("android", response.payload["platform"]?.stringValue)

        bridge.dispose()
    }
}

private class MockDeviceInfoProvider : DeviceInfoProvider {
    var lastVibratePattern: List<Int>? = null

    override fun getDeviceInfo() = DeviceInfo(
        platform = "android",
        osVersion = "14",
        model = "Pixel 8",
        appVersion = "1.0.0"
    )

    override fun getBatteryInfo() = BatteryInfo(level = 85, isCharging = true)

    override fun getScreenInfo() = ScreenInfo(
        width = 390.0,
        height = 844.0,
        scale = 3.0,
        orientation = "portrait"
    )

    override fun vibrate(pattern: List<Int>) {
        lastVibratePattern = pattern
    }

    override suspend fun capturePhoto(quality: Double, camera: String): CapturePhotoResult {
        return CapturePhotoResult(imageBase64 = "mockBase64", width = 1920, height = 1080)
    }

    override suspend fun getLocation(): LocationInfo {
        return LocationInfo(latitude = 37.5665, longitude = 126.978, accuracy = 10.0)
    }

    override suspend fun authenticate(reason: String): AuthenticateResult {
        return AuthenticateResult(success = true)
    }
}
